trezor/trezor-suite
# Message system for suite-native

### TODO

- [x] after merge, warn all developers that they should `rm -rf packages/message-system` and `yarn build:libs`
- [x] tell others, especially @hynek-jina and @trezor/qa  that if they block remote config, it will "fetch"  ⬇️ 
![Screenshot 2023-06-05 at 17 13 26](https://github.com/trezor/trezor-suite/assets/33235762/19b3f952-9e9a-4ccf-bbd0-962f5b44b5bc)

## Description

There is no new UI in this PR, but message system config is being fetched with local fallback, stored in redux state and valid messages based on conditions from config are selected in the store.

Most of this PR is necessary refactoring (moving code around), so I would strongly recommend to watch commit after commit:

- [chore(message-system): move redux part of message-system to suite-common to prepare for suite-native usage](https://github.com/trezor/trezor-suite/pull/7936/commits/880a3ef826d6b8dc0566be220289e8f95bbae316)
  - just rewriting message-system redux part to redux-toolkit and moving to new suite-common/message-system package
- [chore(message-system): jws public key as param for message system init to be able to pass it from mobile app](https://github.com/trezor/trezor-suite/pull/7936/commits/26386d01603d67587b065820e26d6422c13f2aa8)
  - public key is public, so we don't need to read it from secrets... 
  - there is no process.env.JWS_PUBLIC_KEY in suite-native, we need to read the environment variables using react-native-config there
- [feat(message-system): introduce message-system in mobile app](https://github.com/trezor/trezor-suite/pull/7936/commits/9d262122c433fb9597a146f2dabb782c1d6c87a8)
  - just added messageSystem to suite-native redux store
- [chore(suite): move env util from suite to trezor package to make it reusable for native](https://github.com/trezor/trezor-suite/pull/7936/commits/e7fc9c7502f573d2c7a41b81a3951136cec58288)
  - just refactoring
  - `getEnvironment` util will be used for message system validation shared between all environments.
  - Define `Environment` type globally in env-utils instead of analytics where it makes less sense.
- [chore(message-system): move message system utils to common](https://github.com/trezor/trezor-suite/pull/7936/commits/66b0e1ef51e89c84d3a69a14475b46fdb7d1187b)
  - refactoring to make those utils reusable with suite-native.
  - Sorting valid messages to lists of categories extracted to new util `categorizeMessages` to be reusable in suite-native middleware.
- [chore(message-system): delete packages/message-system completely](https://github.com/trezor/trezor-suite/pull/7936/commits/5c1dd37027b010e5e5b74c49e291d2021cc950ea)
  - again refactoring
  - Move all message-system logic to suite-common to prepare for sharing local signed config with suite-native.
  - Everything on one place helps deduplicate VERSION and JWS_SIGN_ALGORITHM constants.
  - Cleaning of unused constants.
  - controversion? [message system types](https://github.com/trezor/trezor-suite/blob/5c1dd37027b010e5e5b74c49e291d2021cc950ea/suite-common/suite-types/src/messageSystem.ts) generated from schema no more gitignored - they are not dynamic, almost never change (only when we change the schema)
- [chore(message-system): store local config in TS file so it can be imported](https://github.com/trezor/trezor-suite/pull/7936/commits/ef9341bbd05087931cf3814b8ae754528d7bc1ee)
  - The local file can't be simply fetched in React Native but we can wrap it in typescript file and dynamically import it instead.
  - maybe controversial? but I've tried different approaches including expo-file-system
- [chore(suite): add suite version and commit hash to env utils](https://github.com/trezor/trezor-suite/pull/7936/commits/56555fe475749985a785eacc5622fbfbea115cdc)
  - refactoring needed for the last commit
- [feat(message-system): selecting valid messages for suite-native](https://github.com/trezor/trezor-suite/pull/7936/commits/7768e5daf96ad69bce5a42d630807be2e464bb33)
  - use everything prepared in previous commits and implement selecting valid messages (based on conditions defined in the config file like app version, commit hash, ...)


### How to test it locally?
1. add some message to config.v1.json file like:
```json
          {
            "conditions": [
                {
                    "environment": {
                        "desktop": "!",
                        "mobile": "*",
                        "web": "!"
                    }
                }
            ],
            "message": {
                "id": "a99d0ad4-3bab-4660-9352-fh1d65181111",
                "priority": 1,
                "dismissible": true,
                "variant": "info",
                "category": "banner",
                "content": {
                    "en-GB": "Hello Suite Lite!",
                    "en": "Hello Suite Lite!",
                    "es": "Hello Suite Lite!",
                    "cs": "Hello Suite Lite!",
                    "ru": "Hello Suite Lite!",
                    "ja": "Hello Suite Lite!"
                }
            }
        },
```
1. increase sequence number in the config file (you need to do it for every change, same or lower sequence is ignored)
1. sign the config by calling `yarn run message-system-sign-config` script
1. `"mobile": "*",` means "any version of Suite Lite" but the version has to be defined, so set VERSION in [.env.debug](https://github.com/trezor/trezor-suite/blob/feat/message-system-common/suite-native/app/.env.debug) to let's say `1`
![image](https://github.com/trezor/trezor-suite/assets/3729633/dfd7927d-d0f7-43cf-8dcb-ea440bc03230)
1. run the App, check the state 🎉 


## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/7274

## Screenshots:
![image](https://github.com/trezor/trezor-suite/assets/3729633/3f7c3a16-8cca-41c4-a0d6-1044885245e7)
